pkg/scheduler/framework/plugins/nodevolumelimits/non_csi.go

```diff
@@ -195,18 +196,51 @@ func (pl *nonCSILimits) Name() string {
 	return pl.name
 }
 
+const (
+	Name              = "NonCSILimits"
+	preFilterStateKey = "PreFilter" + Name
+)
+
+type preFilterState struct {
+	newVolumes sets.String
+}
+
+// Clone just returns the same state because it is not affected by pod additions or deletions.
+func (s *preFilterState) Clone() framework.StateData {
+	return s
+}
+
+func (pl *nonCSILimits) PreFilter(ctx context.Context, state *framework.CycleState, pod *v1.Pod) *framework.Status {
+	newVolumes := make(sets.String)
+	if err := pl.filterVolumes(pod.Spec.Volumes, pod.Namespace, newVolumes); err != nil {
+		return framework.AsStatus(err)
+	}
+	pfState := &preFilterState{newVolumes: newVolumes}
+	state.Write(preFilterStateKey, pfState)
+	return nil
+}
+
+func (pl *nonCSILimits) PreFilterExtensions() framework.PreFilterExtensions {
+	return nil
+}
+
 // Filter invoked at the filter extension point.
-func (pl *nonCSILimits) Filter(ctx context.Context, _ *framework.CycleState, pod *v1.Pod, nodeInfo *framework.NodeInfo) *framework.Status {
+func (pl *nonCSILimits) Filter(ctx context.Context, state *framework.CycleState, pod *v1.Pod, nodeInfo *framework.NodeInfo) *framework.Status {
 	// If a pod doesn't have any volume attached to it, the predicate will always be true.
 	// Thus we make a fast path for it, to avoid unnecessary computations in this case.
 	if len(pod.Spec.Volumes) == 0 {
 		return nil
 	}
 
-	newVolumes := make(sets.String)
-	if err := pl.filterVolumes(pod.Spec.Volumes, pod.Namespace, newVolumes); err != nil {
-		return framework.AsStatus(err)
+	s, err := getPreFilterState(state)
+	if err != nil {
```

Comment:
if there's an error getting the preFilter state, do we want to just return at that point? I think I see the reasoning here of falling back to just calling `pl.filterVolumes` again, but we shouldn't ignore that there was an error in the PreFilter stage

Reply:
I don't think we should do fallback at all. The error here is something not transitory, it is an indication of something really off, and so we shouldn't mask it.

I think that is why you didn't need to update the unit tests! The unit tests should be updated to call PreFilter.

Reply:
OK, so I change it to return the error when fail to get preFilter state. thanks!

Reply:
ok, so we indeed need to continue to fallback to the old logic here for now.

You also need to call PreFilter in the unit tests and fix the the other unit tests that are now failing because PreFilter is now part of the default config.

Reply:
Yes, sir. 👍 
I also fixed the test.